File: 1019_Next_Greater_Node_In_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        vector<int> v;
        stack<int> s;
        for (ListNode * h = head; h; h = h->next) {
            v.push_back(h->val);
        }
        for (int i = v.size() - 1; i >= 0; --i) {
            int t = v[i];
            while (!s.empty() && s.top() <= t)
                s.pop();
            v[i] = s.empty() ? 0 : s.top();
            s.push(t);
        }
        return v;
    }
};

File: 1019_Next_Greater_Node_In_Linked_List.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn next_larger_nodes(head: Option<Box<ListNode>>) -> Vec<i32> {
        let mut v = Vec::<i32>::new();
        let mut s = Vec::<i32>::new();
        let mut h = &head;
        while let Some(n) = h {
            v.push(n.val);
            h = &n.next;
        }
        for i in (0..v.len()).rev() {
            let t = v[i];
            while *s.last().unwrap_or(&2_000_000_000) <= t {
                s.pop();
            }
            v[i] = *s.last().unwrap_or(&0);
            s.push(t);
        }
        v
    }
}